src/main/kotlin/interactor/GetCityHasCheapestInternetConnectionInteractor.kt

```diff
@@ -2,14 +2,55 @@ package interactor
 
 import model.CityEntity
 
-class GetCityHasCheapestInternetConnectionInteractor(
-    private val dataSource: CostOfLivingDataSource,
-) {
 
-    fun execute(): CityEntity{
-        throw Throwable("Not Implemented yet")
+
+class GetCityHasCheapestInternetConnectionInteractor{
+
+
+
+
+    fun execute(cities:List<CityEntity>): CityEntity? {
+        return if (cities.isNotEmpty()) {
+            val filter = cities.filter { excludeNullSalariesAndLowQualityDataAndNullInternetPrice(it) }
+            if (filter.isEmpty()){
+                throw Throwable("List that you entered all of it has bad quality data or null average salary or null internet price")
+            }
+            val result = filter
+                .sortedBy { cityEntity -> (cityEntity.servicesPrices.internet60MbpsOrMoreUnlimitedDataCableAdsl
+                    ?.div(cityEntity.averageMonthlyNetSalaryAfterTax!!))?.times(
+                        100) }.first()
```

Comment:
You could declare a function to do that logic, this **execute** function is just responsible of returning the desired city.
```suggestion
                val result = filter.sortedBy(::findTheRatioOfInternetPriceToSalaryAverage).first()
```

And the **findTheRatioOfInternetPriceToSalaryAverage** be:
```suggestion
private fun findTheRatioOfInternetPriceToSalaryAverage(city: CityEntity) {
        (city.servicesPrices.internet60MbpsOrMoreUnlimitedDataCableAdsl
        ?.div(city.averageMonthlyNetSalaryAfterTax!!))
        ?.times(100)
    }
```